### ISSUE TITLE
When queueing a callable for asap executing should you be able to cancel it?

### ISSUE THREAD

**Issue by WyriHaximus** [async-interop/event-loop#7]
Just like one off and periodic timers you might want to cancel it at some point

**Comment by AndrewCarterUK** [comment]
I'd say so. It's still scheduled and I can't see a reason for not providing the option to remove it.

**Comment by trowski** [comment]
`queue()` should not return a watcher. It should just execute the function next chance it gets with no going back. Essentially you should think of it as the same way as actually executing the function. There's no need to add the performance overhead to create a watcher that will never be used. There is also overhead added to the scheduler if functions can be unscheduled.

**Comment by AndrewCarterUK** [comment]
@trowski - If the purpose is to run it as soon as possible with minimal overhead - wouldn't that code just be run directly rather than added to the loop?

Are there any good examples of when `queue()` would be used? I'm struggling to visualise a good use case.

(edit: Reworded on first paragraph)

**Comment by trowski** [comment]
Icicle uses `queue()` to ensure callbacks on awaitables are always executed asychronously. That is, when calling `then()` or `done()` on an awaitable, the callback is never executed until after the scope calling the awaitable method has been left.

`queue()` is also used to avoid blowing up the call stack when scheduling coroutines. If a coroutine were to have several awaitables in a row that were immediately resolved (with synchronous callbacks), you can easily blow up the call stack.

Generally users of async libraries will have no reason to use `queue()` because they should be using other tools like awaitables (promises) or coroutines to write async code, they shouldn't be scheduling callbacks themselves. However, the event loop needs to have some way of simply deferring execution for these tools to be built on.

**Comment by AndrewCarterUK** [comment]
That all makes sense. None of the use cases given there would require a watcher object - so unless anyone else can think of a use case I'm happy with  `queue()` not returning a watcher. The only reason I can think of in favour is consistency with other methods (edit: other callback registering methods).

**Comment by mbonneau** [comment]
I would like to be able to cancel.

In [RxPHP](https://github.com/ReactiveX/RxPHP) the scheduler needs to be able to schedule items to run as soon as possible, they need to be ordered in the same order they the were scheduled, and they also need to be cancelable.

It is possible (and happens in many Rx operators) that scheduled items will need to cancel scheduled items that are queued behind it.

**Comment by trowski** [comment]
Another watcher could be added for scheduling a callback that could be cancelled but otherwise executed as soon as possible or when it makes sense. Icicle offers [Immediates](https://icicle.io/docs/api/Loop/Watcher.Immediate/) for this purpose. However there needs to be a performant way of deferring function execution. Icicle uses `queue()` extensively to ensure execution is deferred. Adding unnecessary overhead would be very detrimental.

The specification of `queue()` should include that functions MUST be executed in the order given. Icicle, React, and Amp all do this with their equivalent functions already.

RxPHP could build it's own scheduler that uses `queue()` internally when a function is ready to be executed. The scheduler could provide a method to cancel a scheduled function, only calling `queue()` when it is certain the function should be executed. If you're using React, then I assume this sort of thing has already been done since `nextTick()` and `futureTick()` offer no chance to cancel.

**Comment by mbonneau** [comment]
The cancellation feature can be accomplished through other means. And perhaps it would be best to leave this to the user to figure that stuff out as it isn't always needed. This way you do get the best performance.

**Comment by kelunik** [comment]
We have the possibility in Amp. I don't think it's too much overhead.

**Comment by trowski** [comment]
@kelunik If we were just incrementing an integer and returning it I'd be inclined to agree. Allocating and destroying an object for every function deferment is too much overhead.

**Comment by AndrewCarterUK** [comment]
I'll do some benchmarks later to see how significant the overhead is. Will post scripts and results here when done.

**Comment by kelunik** [comment]
@trowski Better increment a string, otherwise you might run into overflows.

**Comment by trowski** [comment]
I suppose you could run into overflow issues if you were running PHP on a 32-bit platform. With phones and other non-traditional computers that's still a high possibility. Whatever, increment an int or string, not much difference.

**Comment by bwoebi** [comment]
I don't think there's a big problem with making immediates (or whatever it's called ;-)) cancellable … though it probably depends a lot on the library around the loop how much overhead it really is.

Amp (I've measured Aerys actually) isn't calling `immediately()` very often, hence there it is rather not an issue. … So, please report how much of overhead it really if for you.

**Comment by kelunik** [comment]
@bwoebi Aerys isn't using that many coroutines, it uses mostly I/O watchers, so might not be the best source to measure.

**Comment by trowski** [comment]
Icicle uses coroutines for virtually everything, so `queue()` is called quite often.

**Comment by bwoebi** [comment]
Hehe, the trick is rather that we're not always scheduling a new tick, but only sometimes: https://github.com/amphp/amp/blob/master/lib/functions.php#L786-L793 (which is a significant perf difference too)

**Comment by trowski** [comment]
I like the consistency offered by making resolution async because resolving a promise will not allow another context to start executing until the current scope is left.

**Comment by bwoebi** [comment]
Well, that doesn't make any difference actually in code (except possibly smaller stack traces) … or am I missing anything?

**Comment by trowski** [comment]
It makes the code much easier to reason about. If a method resolves a promise and modifies the state of the object, synchronous resolution can allow other methods to be called on the same object before the method resolving the promise has finished. You certainly can code around this, but it's a pain and unintuitive.

In the example below, if resolving the deferred causes doSomethingAsync to be called again, the same deferred will be returned as the previous call because the property hasn't been set to null yet.

``` php
class Example
{
    function doSomethingAsync()
    {
        if ($this->deferred) {
            return $this->deferred;
        }

        $this->deferred = new Deferred();

        // Start a watcher in the loop to do something async.
        $this->loop->onReadable($this->stream, function () {
            $this->deferred->resolve($result);
            $this->deferred = null;
        });

        return $this->deferred;
    }
}
```

**Comment by kelunik** [comment]
@trowski That code is not affected, because I/O watchers are only executed in the next tick. It's just about coroutine resolution and doesn't change how the code works.

**Comment by bwoebi** [comment]
I'd say you're describing an anti-pattern.

IMO, you should have your scope cleaned up before you are actually doing calls into the wild - at least Promises shouldn't be tied to the reactor, the immediately is merely a way to distribute work (instead of having a single task suck up all the resources) [implementation detail!].

If you assume a reactor inside promises, that's your thing, but the standardized Promises (if we'll ever have them), shouldn't rely on its existence.

**Comment by trowski** [comment]
@kelunik Resolving the promise will cause the coroutine waiting on the promise to continue. That coroutine might cause `doSomethingAsync()` to be called again.

**Comment by trowski** [comment]
@bwoebi It would probably be wise to set the deferred to null first, and something I could certainly clean up in Icicle. I just worried that others writing code would get caught in the same trap since it's easy to forgot that another context might call your method before it has finished. Even async with coroutines can be hard to reason about. :smile:

**Comment by bwoebi** [comment]
It's a trap and you should avoid it. It's one of the things you need to be aware of when creating a library. But in actual application code (most users - at least when using coroutines), it doesn't matter so much.

**Comment by trowski** [comment]
You make some good points. Only the lower level libs have to deal with these issues, since application code should be using coroutines that don't have those problems. I have some code to modify. :coffee:

**Comment by kelunik** [comment]
Did anyone ever cancel a `defer` / `immediately` / `queue` event yet? If not, the only reason I see is consistency.

**Comment by bwoebi** [comment]
@kelunik no, but someone stated above he's doing it.

Also, there is no harm in supporting it, as we all agreed - thus, closing this issue, except someone has to add anything important.
